storage/filestore/dir.go

```diff
@@ -238,6 +348,136 @@ func (dir *Dir) GarbageCollect(ctx context.Context) (err error) {
 	return nil
 }
 
+const nameBatchSize = 1024
+
+// GetAllNamespaces finds all known namespace IDs in use in local storage. They are not
+// guaranteed to contain any blobs.
+func (dir *Dir) GetAllNamespaces(ctx context.Context) (ids [][]byte, err error) {
+	defer mon.Task()(&ctx)(&err)
+	topBlobDir := dir.blobsdir()
+	openDir, err := os.Open(topBlobDir)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		dirNames, err := openDir.Readdirnames(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if len(dirNames) == 0 {
+			return ids, nil
+		}
+		for _, name := range dirNames {
+			namespace, err := pathEncoding.DecodeString(name)
+			if err != nil {
+				// just an invalid directory entry, and not a namespace. probably
+				// don't need to pass on this error
+				continue
+			}
+			ids = append(ids, namespace)
+		}
+	}
+}
+
+// ForAllKeysInNamespace executes doForEach for each locally stored blob, stored with storage
+// format V1 or greater, in the given namespace. If doForEach returns a non-nil error,
+// ForAllKeysInNamespace will stop iterating and return the error immediately.
+func (dir *Dir) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	namespaceDir := pathEncoding.EncodeToString(namespace)
+	nsDir := filepath.Join(dir.blobsdir(), namespaceDir)
+	openDir, err := os.Open(nsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// job accomplished: there are no blobs in this namespace!
+			return nil
+		}
+		return err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
```

Comment:
Any reason this isn't using `filepath.Walk`?

Reply:
Only that `filepath.Walk` would make the logic considerably harder to follow, in my opinion. It would be necessary to parse the path each iteration to discover which namespace and which key-prefix is being traversed, and use SkipDir for unexpected directories, etc.